### PR TITLE
MLHR-1746 #resolve

### PR DESCRIPTION
**Pull request by pm48**
Created concrete implementation of couchdb input operator.

### CODE REVIEW

File: contrib/src/main/java/com/datatorrent/contrib/couchdb/CouchDbPOJOInputOperator.java [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+      objectClass = Class.forName(outputClass);
+    }
+    catch (ClassNotFoundException ex) {
+      throw new RuntimeException(ex);
+    }
+
+    setterDocId = PojoUtils.createSetter(objectClass, expressionForDocId, String.class);
+    for (int i = 0; i < expressions.size(); i++) {
+      try {
+        setterDoc.add(PojoUtils.createSetter(objectClass, expressions.get(i), Class.forName(fieldType.get(i))));
```
Comment (PramodSSImmaneni) [inline contrib/src/main/java/com/datatorrent/contrib/couchdb/CouchDbPOJOInputOperator.java]: Can you cache the class types

File: contrib/src/main/java/com/datatorrent/contrib/couchdb/CouchDbPOJOInputOperator.java [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+      obj = objectClass.newInstance();
+    }
+    catch (InstantiationException ex) {
+      throw new RuntimeException(ex);
+    }
+    catch (IllegalAccessException ex) {
+      throw new RuntimeException(ex);
+    }
+    setterDocId.set(obj, value.getId());
+    for (int i = 0; i < expressions.size(); i++) {
```
Comment (PramodSSImmaneni) [inline contrib/src/main/java/com/datatorrent/contrib/couchdb/CouchDbPOJOInputOperator.java]: setterDoc size

File: contrib/src/main/java/com/datatorrent/contrib/couchdb/CouchDbPOJOInputOperator.java (hunk 1 above)
Comment (pm48) [inline contrib/src/main/java/com/datatorrent/contrib/couchdb/CouchDbPOJOInputOperator.java]: Why do we need to cache them? They are not being used anywhere else in the operator.

File: contrib/src/main/java/com/datatorrent/contrib/couchdb/CouchDbPOJOInputOperator.java [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+ * Each row is read as a jsonNode and converted into POJO using an ObjectMapper.
+ * The POJO generated will also contain the document Id and revision of the document fetched.
+ * This implementation uses the emitTuples implementation of {@link AbstractCouchDBInputOperator} which emits the complete result
+ * of the ViewQuery every window cycle.
+ * User should provide the viewQuery to fetch the rows from database.
+ **/
+public class CouchDbPOJOInputOperator extends AbstractCouchDBInputOperator<Object>
+{
+  @NotNull
+  private List<String> expressions;
```
Comment (PramodSSImmaneni) [inline contrib/src/main/java/com/datatorrent/contrib/couchdb/CouchDbPOJOInputOperator.java]: Why are there no list of columns here?

File: contrib/src/main/java/com/datatorrent/contrib/couchdb/CouchDbPOJOInputOperator.java [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+   * POJO class which is generated as output from this operator.
+   * Example:
+   * public class TestPOJO{ int intfield; public int getInt(){} public void setInt(){} }
+   * outputClass = TestPOJO
+   * POJOs will be generated on fly in later implementation.
+   */
+  private transient Class<?> objectClass = null;
+  private String outputClass;
+  @NotNull
+  private List<String> fieldType;
```
Comment (PramodSSImmaneni) [inline contrib/src/main/java/com/datatorrent/contrib/couchdb/CouchDbPOJOInputOperator.java]: Can the field types specification not be mandatory. If they are empty the expressions can be assumed to represent field accessors and the field types could be inferred from the POJO definition itself.

File: contrib/src/main/java/com/datatorrent/contrib/couchdb/CouchDbPOJOInputOperator.java [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+ * This implementation uses the emitTuples implementation of {@link AbstractCouchDBInputOperator} which emits the complete result
+ * of the ViewQuery every window cycle.
+ * User should provide the viewQuery to fetch the rows from database.
+ **/
+public class CouchDbPOJOInputOperator extends AbstractCouchDBInputOperator<Object>
+{
+  @NotNull
+  private List<String> expressions;
+  @NotNull
+  private String expressionForDocId;
```
Comment (PramodSSImmaneni) [inline contrib/src/main/java/com/datatorrent/contrib/couchdb/CouchDbPOJOInputOperator.java]: Why can't this be optional

File: contrib/src/main/java/com/datatorrent/contrib/couchdb/CouchDbPOJOInputOperator.java [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+ * of the ViewQuery every window cycle.
+ * User should provide the viewQuery to fetch the rows from database.
+ *
+ */
+public class CouchDbPOJOInputOperator extends AbstractCouchDBInputOperator<Object>
+{
+  //List of expressions set by User. Example:setId(),setName(),Address
+  @NotNull
+  private List<String> expressions;
+  @NotNull
```
Comment (PramodSSImmaneni) [inline contrib/src/main/java/com/datatorrent/contrib/couchdb/CouchDbPOJOInputOperator.java]: This should be optional

File: contrib/src/main/java/com/datatorrent/contrib/couchdb/CouchDbPOJOInputOperator.java [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+  /*
+   * View-query that specifies the couch-db view whose results will be fetched.
+   */
+  @Override
+  public ViewQuery getViewQuery()
+  {
+    return viewQuery;
+  }
+
+  public void setViewQuery(ViewQuery viewQuery)
```
Comment (PramodSSImmaneni) [inline contrib/src/main/java/com/datatorrent/contrib/couchdb/CouchDbPOJOInputOperator.java]: Can ViewQuery be created in UI, if not provide the user a way to supply the query information.

File: contrib/src/main/java/com/datatorrent/contrib/couchdb/CouchDbPOJOInputOperator.java [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+    }
+    catch (IllegalAccessException ex) {
+      throw new RuntimeException(ex);
+    }
+    setterDocId.set(obj, value.getId());
+    for (int i = 0; i < setterDoc.size(); i++) {
+      Class<?> type = fieldType.get(i);
+      if (type.isPrimitive()) {
+        if (type == int.class) {
+          ((SetterInt)setterDoc.get(i)).set(obj, value.getValueAsNode().get(columns.get(i)).getIntValue());
```
Comment (PramodSSImmaneni) [inline contrib/src/main/java/com/datatorrent/contrib/couchdb/CouchDbPOJOInputOperator.java]: Don't repeat getting value as node

**Comment by pm48** [comment]
Made changes as suggested in above coments.

File: contrib/src/main/java/com/datatorrent/contrib/couchdb/CouchDbPOJOInputOperator.java [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+  @NotNull
+  private List<String> expressions;
+  private String expressionForDocId;
+  // List of columns provided by User. Example: id,name,address
+  @NotNull
+  private List<String> columns;
+  @NotNull
+  private String designDocumentName;
+  @NotNull
+  private String viewName;
```
Comment (PramodSSImmaneni) [inline contrib/src/main/java/com/datatorrent/contrib/couchdb/CouchDbPOJOInputOperator.java]: What name is this? Is this the mapping function? How are the mapping and reducing functions specified?

File: contrib/src/main/java/com/datatorrent/contrib/couchdb/CouchDbPOJOInputOperator.java [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+  {
+    this.outputClass = outputClass;
+  }
+
+  @Override
+  public void setup(OperatorContext context)
+  {
+    super.setup(context);
+    setterDoc = new ArrayList<Object>();
+    query = new ViewQuery().designDocId(designDocumentName).viewName(viewName);
```
Comment (PramodSSImmaneni) [inline contrib/src/main/java/com/datatorrent/contrib/couchdb/CouchDbPOJOInputOperator.java]: The view runs on the entire set of documents so it should be pre-created and the view should track newer documents (Do verify this). We need a query to run on the view that will return the documents progressively.

File: contrib/src/main/java/com/datatorrent/contrib/couchdb/CouchDbPOJOInputOperator.java [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+      }
+      else {
+        ((Setter<Object, Object>)setterDoc.get(i)).set(obj, mapper.readValue(val.get(columns.get(i)), type));
+      }
+    }
+
+    //Update start key and skip parameter in case user specified it.It will ultimately be updated to the value in last row's key.
+    if (startKey != null) {
+      startKey = value.getKey();
+      skip = 1;
```
Comment (PramodSSImmaneni) [inline contrib/src/main/java/com/datatorrent/contrib/couchdb/CouchDbPOJOInputOperator.java]: How is this reset?

File: contrib/src/main/java/com/datatorrent/contrib/couchdb/CouchDbPOJOInputOperator.java (hunk 10 above)
Comment (pm48) [inline contrib/src/main/java/com/datatorrent/contrib/couchdb/CouchDbPOJOInputOperator.java]: Yes views track newer documents by setting stale parameter = false on
query.View is updated before query is executed. This ensures that any
documents updated (and persisted to disk) are included in the view.

**Pull request closed without merging**